File: lc-0020/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        auto isPair = [](char l, char r) {
            return (l == '(' && r == ')') || (l == '[' && r == ']') || (l == '{' && r == '}');
        };

        stack<char> t;
        for (auto c : s) {
            if (c == '(' || c == '[' || c == '{') {
                t.push(c);
            } else {
                if (t.empty()) return false;
                char l = t.top();
                t.pop();
                if (!isPair(l, c)) {
                    return false;
                }
            }
        }
        return t.empty();
    }
};

File: lc-0020/solution.rs
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let is_pair = |l, r| {
            match (l, r) {
                ('(', ')') | ('[', ']') | ('{', '}') => true,
                _ => false,
            }
        };

        let mut t = Vec::new();
        for c in s.chars() {
            match c {
                '(' | '[' | '{' => t.push(c),
                _ => {
                    if let Some(l) = t.pop() {
                        if !is_pair(l, c) {
                            return false;
                        }
                    } else {
                        return false;
                    }
                }
            }
        }
        return t.is_empty();
    }
}
